kong/plugins/request-transformer/access.lua

```diff
@@ -212,15 +212,24 @@ local function transform_headers(conf)
 
   -- Add header(s)
   for _, name, value in iter(conf.add.headers) do
-    name = name:lower()
-    if not headers[name] and name ~= HOST then
+    if not headers[name] and name:lower() ~= HOST then
       headers[name] = value
     end
   end
 
   -- Append header(s)
   for _, name, value in iter(conf.append.headers) do
     if name:lower() ~= HOST then
+      if headers[name] ~= nil then -- keep original content, use configd case
+        for header_name, _ in pairs(headers) do
+          if name:lower() == header_name:lower() then
+            local content = headers[header_name]
+            headers[header_name] = nil
+            headers[name] = content
+            break
+          end
+        end
+      end
       headers[name] = append_value(headers[name], value)
     end
   end
```

Comment:
I see, how about this version? I tested it locally and it seems to work well.

What we are doing here is to avoid a O(n) situation when checking the `headers` table for duplicates. Since `get_headers` always return request headers in all lower case, we can do a single lookup instead of loops to avoid this extra looping (especially considering `pairs` is not JITable and known to be slow).
```suggestion
  for _, name, value in iter(conf.append.headers) do
    local name_lc = name:lower()

    if name_lc ~= HOST and name ~= name_lc and headers[name] ~= nil then
      -- keep original content, use configd case
      headers[name] = headers[name_lc]
      headers[name_lc] = nil
    end

    headers[name] = append_value(headers[name], value)
  end
```